torch/csrc/inductor/static_cuda_launcher.cpp

```diff
@@ -0,0 +1,502 @@
+    return nullptr;
+  }
+  // Return a tuple of CUFunction, n_regs, n_spills
+  return Py_BuildValue(
+      "(Kii)", reinterpret_cast<uint64_t>(func), n_regs, n_spills);
+}
+
+template <size_t NUM_ARGS>
+static PyObject* launch_kernel_inner(
+    CUfunction func,
+    int gridX,
+    int gridY,
+    int gridZ,
+    int numWarps,
+    int sharedMemBytes,
+    const char* argTypes,
+    PyObject* varArgs,
+    cudaStream_t cudaStream) {
+  try {
+    // Launch the kernel
+    // Prepare the arguments for the kernel
+    // We allocate 8 bytes per argument on the stack. We then allocate 8 more
+    // bytes to point to each 8 byte slot in argStorage, and pass that array of
+    // pointers to launchKernel.
+    std::array<uint64_t, NUM_ARGS> argStorage = {};
+    std::array<void*, NUM_ARGS> kernelArgs = {};
+    parseKernelArgs(varArgs, argTypes, argStorage, kernelArgs);
+
+    launchKernel(
+        func,
+        gridX,
+        gridY,
+        gridZ,
+        numWarps,
+        sharedMemBytes,
+        kernelArgs,
+        cudaStream);
+  } catch (const std::runtime_error& e) {
+    PyErr_SetString(PyExc_RuntimeError, e.what());
+    return nullptr;
+  }
+  Py_RETURN_NONE;
+}
+#define LAUNCH_KERNEL(N)  \
+  launch_kernel_inner<N>( \
+      func,               \
+      gridX,              \
+      gridY,              \
+      gridZ,              \
+      numWarps,           \
+      sharedMemBytes,     \
+      argTypes,           \
+      varArgs,            \
+      cudaStream)
+
+/**
+ *  Main entrypoint function called at runtime; called like this in python land:
+    launcher(
+      function, # CUfunction returned by load_kernel()
+      grid_x,
+      grid_y,
+      grid_z,
+      num_warps,
+      shared,
+      arg_tys, # e.g. "bO" for (int8_t, uint64_t)
+      args, # tuple of arguments passed to the kernel
+      stream,
+  )
+ *
+ */
+static PyObject* launch_kernel(PyObject* self, PyObject* args) {
+  // Pointer to CUfunction generated by load_kernel()
+  uint64_t func_ptr = 0;
+  int gridX = 0, gridY = 0, gridZ = 0, numWarps = 0, sharedMemBytes = 0;
+  // stream here should be the raw stream gotten from
+  // device_interface.get_raw_stream()
+  uint64_t stream = 0;
+  const char* argTypes = nullptr;
+  PyObject* varArgs = nullptr;
+  // Parse the fixed arguments and the format string
+  if (!PyArg_ParseTuple(
+          args,
+          "KiiiiisOl",
+          &func_ptr,
+          &gridX,
+          &gridY,
+          &gridZ,
+          &numWarps,
+          &sharedMemBytes,
+          &argTypes,
+          &varArgs,
+          &stream)) {
+    return nullptr;
+  }
+  CUfunction func = reinterpret_cast<CUfunction>(func_ptr); // NOLINT
+  cudaStream_t cudaStream = reinterpret_cast<cudaStream_t>(stream); // NOLINT
+  auto num_args = std::strlen(argTypes);
+  // Support up to 10 args by default
+  // We use a templated function here to avoid heap allocations for the
+  // argument.
+  switch (num_args) {
```

Comment:
We can do some profiling to decide the correct number here. @aorenste assured me that this will compile into a jump table, so isn't too slow to branch at kernel runtime (over, say, defining N python functions and branching at compile time instead)

Reply:
Is it possible to have a kernel with 0 parameters? Do we need support for that?

Reply:
we do not - even for compiling a constructor like `zeros` the output buffer gets passed in